torchgeo/trainers/simclr.py

```diff
@@ -0,0 +1,242 @@
+        if num_layers > 2:
+            layers.extend(
+                [(hidden_dim, hidden_dim, nn.BatchNorm1d(hidden_dim), nn.ReLU())]
+                * (num_layers - 2)
+            )
+        layers.append((hidden_dim, output_dim, nn.BatchNorm1d(output_dim), None))
+        super().__init__(layers)
+
+
+class SimCLRTask(LightningModule):  # type: ignore[misc]
+    """SimCLR: a simple framework for contrastive learning of visual representations.
+
+    Reference implementation:
+
+    * https://github.com/google-research/simclr
+
+    If you use this trainer in your research, please cite the following papers:
+
+    * v1: https://arxiv.org/abs/2002.05709
+    * v2: https://arxiv.org/abs/2006.10029
+
+    .. versionadded:: 0.5
+    """
+
+    def __init__(
+        self,
+        model: str = "resnet50",
+        weights: Optional[Union[WeightsEnum, str, bool]] = None,
+        in_channels: int = 3,
+        version: int = 2,
+        layers: int = 3,
+        hidden_dim: Optional[int] = None,
+        output_dim: Optional[int] = None,
+        lr: float = 4.8,
+        weight_decay: float = 1e-4,
+        max_epochs: int = 100,
+        temperature: float = 0.07,
+        memory_bank_size: int = 64000,
+        gather_distributed: bool = False,
+        augmentations: nn.Module = AUG,
+    ) -> None:
+        """Initialize a new SimCLRTask instance.
+
+        Args:
+            model: Name of the timm model to use.
+            weights: Either a weight enum, the string representation of a weight enum,
+                True for ImageNet weights, False or None for random weights,
+                or the path to a saved model state dict.
+            in_channels: Number of input channels to model.
+            version: Version of SimCLR, 1--2.
+            layers: Number of layers in projection head (2 for v1 or 3+ for v2).
+            hidden_dim: Number of hidden dimensions in projection head
+                (defaults to output dimension of model).
+            output_dim: Number of output dimensions in projection head
+                (defaults to output dimension of model).
+            lr: Learning rate (0.3 x batch_size / 256 is recommended).
+            weight_decay: Weight decay coefficient (1e-6 for v1 or 1e-4 for v2).
+            max_epochs: Maximum number of epochs to train for.
+            temperature: Temperature used in NT-Xent loss.
+            memory_bank_size: Size of memory bank (0 for v1 or 64K for v2).
+            gather_distributed: Gather negatives from all GPUs during distributed
+                training (ignored if memory_bank_size > 0).
+            augmentations: Data augmentation.
+        """
+        super().__init__()
+
+        self.save_hyperparameters(ignore=["augmentations"])
+        self.augmentations = augmentations
+
+        # Create backbone
+        self.backbone = timm.create_model(
+            model, in_chans=in_channels, pretrained=weights is True
```

Comment:
Should we set num_classes=0 which would then return the backbone features before the classification layer?

Reply:
Is that what that does? I can't even find documentation for anything in timm. Also not sure if I know TF well enough to determine whether or not they replace the last layer or not.

Reply:
Any idea if SimCLR does this too? I'll keep digging

Reply:
What I'm saying is you NEED to set num_classes=0 otherwise num_classes defaults to 1000 for imagenet and will return classifier logits. By setting num_classes=0 it will return a model without the classifier head which is what we want.

Reply:
Lightly example also does this. Let me make the change.